### DISCUSSION TITLE
Memorials: how to get the normal readings.

### DISCUSSION THREAD

**Discussion by pvanthony** [romcal/romcal#321] [Q&A] [answered]
Trying to get daily readings on the website. The issue that we are facing is on days with Memorial. We have only a few days which does the reading for Memorial. The other memorial days we do the normal readings. 

In version 2, we used the "prioritized: true" item in the array to do a check then select the correct array item for the reading.

Is this the correct way of doing this? Is there a better way to do this in version 3?

By the way, from my understanding, these readings are the same all over the world. Am I correct?

In version 2, we did a country file and compiled it. Here is the country file. Need advice on what is the correct way to get the daily readings.

```
import * as Locales from '@romcal/lib/locales';
import { LiturgicalColors } from '@romcal/constants/liturgical-colors/liturgical-colors.enum';
import { LiturgicalDayInput } from '@romcal/models/liturgical-day/liturgical-day.types';
import { RomcalConfig, RomcalConfigInCalendarDef } from '@romcal/models/config/config.model';
import { Ranks } from '@romcal/constants/ranks/ranks.enum';
import { Titles } from '@romcal/constants/titles/titles.enum';

const defaultConfig: RomcalConfigInCalendarDef | undefined = undefined;

const dates = async (config: RomcalConfig): Promise<Array<LiturgicalDayInput>> => {
  const year = config.year;
  const _dates: Array<LiturgicalDayInput> = [
    {
      key: 'martha_of_bethany_mary_of_bethany_and_lazarus_of_bethany',
      extend: true,
      prioritized: true,
    },
    {
      key: 'our_lady_of_sorrows',
      extend: true,
      prioritized: true,
    },
    {
      key: 'therese_of_the_child_jesus_and_the_holy_face_of_lisieux_virgin',
      extend: true,
      prioritized: true,
    },
    {
      key: 'guardian_angels',
      extend: true,
      prioritized: true,
    },
    {
      key: 'francis_xavier_priest',
      extend: true,
      prioritized: true,
    },
    {
      key: 'barnabas_apostle',
      extend: true,
      prioritized: true,
    },
    {
      key: 'passion_of_saint_john_the_baptist',
      extend: true,
      prioritized: true,
    },
  ];

  // Get localized liturgical day names
  return await Locales.localizeDates(_dates);
};

export { dates, defaultConfig };
```

**Comment by tukusejssirs** [comment]
I think the `prioritized` was misused in romcal v2 by you.

The first thing that comes to my mind is that if you are sure that there are only 7 memorials, you could you a simple `switch`. I don’t think it is a perfect solution, but maybe @emagnier can think of a better way.

**Comment by emagnier** [comment, marked as answer]
In v3, the `prioritized` option has been removed (all the management and precedence rules between liturgical days has been redesigned). 
Until `v2`, when you set `prioritized: true` on an item, all other items are ignored and only the "prioritized" item is output. But this can cause issues if this memorial is falling in a Sunday (in that case, Sunday will be ignored), or if any possible feasts or solemnities that could also occur the same day.

You created a local calendar (what you called a "country file") and this is the right way to go.

From your parish or local community, if you decided to always celebrate a specific memorial that is proposed as an optional memorial (and also eventually to use its readings, but as you know readings are not managed yet in romcal), I suggest elevating to a higher rank the corresponding memorial in your local calendar file (`precedence: Precedences.ProperMemorial_11b`), so prioritization of items is always managed correctly by romcal (i.e. if the memorial is falling a Sunday, or if there is another item with a higher rank the same day).

For example below, a local calendar where we decided to celebrate John XXIII and John Paul II instead of the default weekday. These definitions are already defined in the General Roman Calendar, but from this calendar, their precedence are redefined to be elevated.
```ts
import { Precedences } from '../constants/precedences';
import { CalendarDef } from '../models/calendar-def';
import { Inputs } from '../types/calendar-def';
import { France } from './france';

export class LocalCalendar extends CalendarDef {
  // If your local calendar is a derived from a country calendar
  // you can define a parent calendar.
  // E.g. if your local calendar is used in France:
  parentCalendar = France;

  inputs: Inputs = {
    john_xxiii_pope: {
      precedence: Precedences.ProperMemorial_11b,
    },

    john_paul_ii_pope: {
      precedence: Precedences.ProperMemorial_11b,
    },
  };
}
```

Finally, this still doesn't solve the cases where we decided to take the readings from the weekday, or the readings from the memorial.
But, for example, you could create an array of keys of memorials where you want to use the proper readings.

```js
const properReadingsForMemorials = [
  'martha_of_bethany_mary_of_bethany_and_lazarus_of_bethany',
  'our_lady_of_sorrows',
  'therese_of_the_child_jesus_and_the_holy_face_of_lisieux_virgin',
  'guardian_angels',
  'francis_xavier_priest',
  'barnabas_apostle',
  'anthony_of_padua_priest',
  'passion_of_saint_john_the_baptist',
];

if (currentItem.rank === 'MEMORIAL') {
  if (properReadingsForMemorials.includes(currentItem.key)) {
    // This current LiturgicalDay have proper readings.
  } else {
    // This current LiturgicalDay use the readings of the weekday.
    const weekdayItem = currentItem.weekday;
    const weekdayKey = weekdayItem ? weekdayItem.key : undefined;
  }
} else {
  // This is not a memorial: the readings are those of the current LiturgicalDay
}
```

> By the way, from my understanding, these readings are the same all over the world. Am I correct?

Yes they are. 
However, the exact verse references might be slightly different depending on the translation.

**Reply by pvanthony** [reply]
Thank you very much for the solution.

 I will create a Singapore calendar for our own testing. When we have tested it completely, I will post it here.

Also thank you for example codes. Very helpful. Saved a lot of time.

Once again thank you both for helping. I do appreciate it.

**Reply by emagnier** [reply]
Happy to help ;)
I've updated a little the last code example in my answer, to show better the specific management of memorials, and all other rank types that are not memorials.

**Reply by pvanthony** [reply]
With your code, I do not have to create the Singapore calendar file. It just works. Great!!!

**Reply by tukusejssirs** [reply]
> With your code, I do not have to create the Singapore calendar file. It just works. Great!!!

You don’t need to create `singapore.ts` calendar unless there are some modifications in comparison with the GRC (like additional celebrations, rank raising/lowering, transfering a celebration to another date or even celebration omissions). However, if there are any changes, I highly recommend to you to create `singapore.ts` or provides us the changes and let us create that calendar for you. :wink:

**Comment by pvanthony** [comment]
Please help to check 2nd of January weekday name item. Getting the following.

Memorial:  2021-01-02 Saints Basil the Great and Gregory Nazianzen, Bishops and Doctors of the Church 
Other Reading:  January 2 <--- seems wrong.

The code used was as follows:
```
            for(day in result){
                if(result[day][0].rank == "MEMORIAL"){
                    console.log("Memorial: ", result[day][0].date, result[day][0].name, result[day][0].colorNames)
                    console.log("Other Reading: ", result[day][0].weekday.name)
                }
            }
```

**Reply by tukusejssirs** [reply]
@pvanthony, I don’t really understand you problem here.

For me, the following code produces the output bellow:

```js
// Code
console.log("Memorial: ", result['2021-01-02'][0].date, result['2021-01-02'][0].name, result['2021-01-02'][0].colorNames)
console.log("Other Reading: ", result['2021-01-02'][0].weekday.name)

// Output
Memorial:  2021-01-02 Saints Basil the Great and Gregory Nazianzen, Bishops and Doctors of the Church 
Array [ "white" ]
Other Reading:  January 2
```

Are you missing something? Or do you want to reformat the date (`January 2` → `2 January`)?

**Reply by pvanthony** [reply]
My mistake. There is such a reading as January 2.

My bad.
